File: C/cyclesort/qwen-qwq-32b.c
void cycleSort(int arr[], int n) {

{

    for (int cycleStart = 0; cycleStart < n; cycleStart++) {

        int item = arr[cycleStart];

        int pos = cycleStart;

        for (int i = cycleStart + 1; i < n; i++)

            if (arr[i] < item)

                pos++;

 

        // Handle duplicates.

        while (item == arr[pos]

            pos++;

 

        if (pos != cycleStart) {

            swap(&arr[pos], &item);

        }

 

        // Now, continue the cycle.

        while (pos != cycleStart) {

            pos = cycleStart;

            for (int i = cycleStart + 1; i < n; i++)

                if (arr[i] < item)

                    pos++;

                else if (arr[i] == item)

                    pos++;

 

            while (item == arr[pos]

                pos++;

 

            if (pos != cycleStart) {

                swap(&arr[pos], &item);

            }

        }

    }

}

File: Python/cyclesort/qwen-qwq-32b.py


def cycleSort(array):
    writes = 0
    n = len(array)
    for cycleStart in range(n - 1):
        item = array[cycleStart]
        pos = cycleStart
        for i in range(cycleStart + 1, n):
            if array[i] < item:
                pos += 1
        if pos == cycleStart:
            continue
        while item == array[pos]:
            pos += 1
        array[pos], item = item, array[pos]
        writes +=1
        while pos != cycleStart:
            pos = cycleStart
            for i in range(cycleStart +1, n):
                if array[i] < item:
                    pos +=1
            while item == array[pos]:
                pos +=1
            array[pos], item = item, array[pos]
            writes +=1
    return writes

import sys
def main():
    input_path = "inputs.txt" if len(sys.argv) < 2 else sys.argv[1]

    try:
        with open(input_path, 'r') as f:
            arr = [int(line.strip()) for line in f if line.strip()]
    except FileNotFoundError:
        return

    cycleSort(arr)


if __name__ == "__main__":
    main()
